src/core/search/compressed_list.cc

```diff
@@ -0,0 +1,210 @@
+  absl::Span<const uint8_t> diffs_tail{diffs_}, entry_span{};
+
+  while (entry < value && !diffs_tail.empty()) {
+    auto [diff, read] = ReadVarLen(diffs_tail);
+
+    prev_entry = entry;
+    entry += diff;
+
+    entry_span = diffs_tail.subspan(0, read);
+    diffs_tail.remove_prefix(read);
+  }
+
+  return EntryLocation{entry, prev_entry, entry_span};
+}
+
+// Insert has linear complexity. It tries to find between which two entries A and B the new value V
+// needs to be inserted. Then it computes the differences dif1 = V - A and diff2 = B - V that need
+// to be stored to encode the triple A V B. Those are stored where diff0 = B - A was previously
+// stored, possibly extending the vector
+void CompressedList::Insert(uint32_t value) {
+  auto bound = LowerBound(value);
+
+  // At least one element was read and it's equal to value: return to avoid duplicate
+  if (bound.value == value && !bound.diff_span.empty())
+    return;
+
+  // We're inserting below unconditionally
+  size_++;
+
+  // Value is bigger than any other (or list is empty): append required diff at the end
+  if (value > bound.value || bound.diff_span.empty()) {
+    PushBackDiff(value - bound.value);
+    return;
+  }
+
+  // Now the list certainly contains the bound B > V and possibly A < V (or 0 by default),
+  // so we need to encode both differences diff1 and diff2
+  DCHECK_GT(bound.value, value);
+  DCHECK_LT(bound.prev_value, value);
+
+  // Compute and encode new diff1 and diff2 into buf1 and buf2 respectivaly
+  array<uint8_t, 16> buf1, buf2;
+  auto diff1_span = WriteVarLen(value - bound.prev_value, absl::MakeSpan(buf1));
+  auto diff2_span = WriteVarLen(bound.value - value, absl::MakeSpan(buf2));
+
+  // Extend the location where diff0 is stored with optional zeros before overwriting it
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  size_t required_len = diff1_span.size() + diff2_span.size();
+  DCHECK_LE(bound.diff_span.size(), required_len);  // It can't shrink for sure
+  diffs_.insert(diffs_.begin() + diff_offset, required_len - bound.diff_span.size(), 0u);
+
+  // Now overwrite diff0 and 0s with the two new differences
+  copy(diff1_span.begin(), diff1_span.end(), diffs_.begin() + diff_offset);
+  copy(diff2_span.begin(), diff2_span.end(), diffs_.begin() + diff_offset + diff1_span.size());
+}
+
+// Remove has linear complexity. It tries to find the element V and its neighbors A and B,
+// which are encoded as diff1 = V - A and diff2 = B - V. Adjacently stored diff1 and diff2
+// need to be replaced with diff3 = diff1 + diff2s
+void CompressedList::Remove(uint32_t value) {
+  auto bound = LowerBound(value);
+
+  // Nothing was read or the element was not found
+  if (bound.diff_span.empty() || bound.value != value)
+    return;
+
+  // We're removing below unconditionally
+  size_--;
+
+  // Calculate offset where values diff is stored and determine diffs tail
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
```

Comment:
Isn't this necessarily `bound.diff_span.size()`?
If so, maybe add a `DCHECK()` or just use that instead, up to you

Reply:
no, diff_span covers only the small part that encodes the difference of value, not the whole prefix